### PR TITLE
Publishing timeout to log at WARN and indicate pending nodes

### PR DESCRIPTION
**Pull request by bleskes**
When the master publishes a new cluster state it waits (by default) for up to 30s for all nodes to respond. If not it continues to process other pending tasks. At the moment, this timeout is logged under DEBUG but it typically represent a serious issue with one or more of the nodes. We should log it in WARN and give the nodes that failed to respond in a timely fashion

We also remove the ClusterStatePublishResponseHandler interface because it is an unneeded abstraction, at least for now.

### CODE REVIEW

**Comment by martijnvg** [comment]
LGTM

File: src/test/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandlerTests.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        threads = new Thread[secondRound];
+        barrier = new CyclicBarrier(secondRound);
+
+        for (int i = 0; i < threads.length; i++) {
+            threads[i] = new Thread(new PublishResponder(randomBoolean(), allNodes[firstRound + i], barrier, logger, handler));
+            threads[i].start();
+        }
+        // wait on the threads to finish
+        for (Thread t : threads) {
+            t.join(30000);
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandlerTests.java]: can this be a join without a timeout?

File: src/test/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandlerTests.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        CyclicBarrier barrier = new CyclicBarrier(firstRound);
+        Set<DiscoveryNode> completedNodes = new HashSet<>();
+        for (int i = 0; i < threads.length; i++) {
+            completedNodes.add(allNodes[i]);
+            threads[i] = new Thread(new PublishResponder(randomBoolean(), allNodes[i], barrier, logger, handler));
+            threads[i].start();
+        }
+        // wait on the threads to finish
+        for (Thread t : threads) {
+            t.join(30000);
```
Comment (s1monw) [inline src/test/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandlerTests.java]: can this be a join without a timeout?

File: src/main/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandler.java [hunk 3]
```diff
@@ -20,41 +20,71 @@
...
-        return latch.await(timeout.millis(), TimeUnit.MILLISECONDS);
+        boolean success = latch.await(timeout.millis(), TimeUnit.MILLISECONDS);
+        assert !success || pendingNodes.isEmpty() : "response count reached 0 but still waiting for some nodes";
+        return success;
+    }
+
+    /**
+     * returns a list of nodes which didn't respond yet
+     */
+    public DiscoveryNode[] pendingNodes() {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandler.java]: can't we just use Iterable<DicsoceryNode> and make it unmodifiable?

File: src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java [hunk 4]
```diff
@@ -152,7 +159,11 @@ public void handleException(TransportException exp) {
             try {
                 timedOutWaitingForNodes.set(!publishResponseHandler.awaitAllNodes(publishTimeout));
                 if (timedOutWaitingForNodes.get()) {
-                    logger.debug("timed out waiting for all nodes to process published state [{}] (timeout [{}])", clusterState.version(), publishTimeout);
+                    DiscoveryNode[] pendingNodes = publishResponseHandler.pendingNodes();
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: hmm can we simply just have a `#numPendingNodes()` or `#hasPendingNodes()`

File: src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java [hunk 5]
```diff
@@ -355,7 +363,11 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                 try {
                     boolean awaited = publishResponseHandler.awaitAllNodes(publishTimeout);
                     if (!awaited) {
-                        logger.debug("awaiting all nodes to process published state {} timed out, timeout {}", clusterState.version(), publishTimeout);
+                        DiscoveryNode[] pendingNodes = publishResponseHandler.pendingNodes();
+                        // everyone may have just responded
+                        if (pendingNodes.length > 0) {
```
Comment (s1monw) [inline src/main/java/org/elasticsearch/discovery/local/LocalDiscovery.java]: seem my comment below about `#hasPendingNodes`

**Comment by s1monw** [comment]
LGTM in general left some cosmetic comments

File: src/main/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandler.java (hunk 3 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandler.java]: I can expose the iterable of the Set but then it just exposes the semantics of what an iterator means based on the underlying set implementation. I felt it's simpler this way.

File: src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 4 above)
Comment (bleskes) [inline src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: Not sure I follow? my point here was to log the lagging nodes, so one can go to them and research what's going on.

File: src/test/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandlerTests.java (hunk 1 above)
Comment (bleskes) [inline src/test/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandlerTests.java]: will do.

File: src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java (hunk 4 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java]: nevermind I didn't see the logging statement :)

File: src/main/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandler.java (hunk 3 above)
Comment (s1monw) [inline src/main/java/org/elasticsearch/discovery/BlockingClusterStatePublishResponseHandler.java]: fair enough

**Pull request closed without merging**